### PR TITLE
Products: Access to related up sell cross sell product fields

### PR DESCRIPTION
**Pull request by VoronoyAlexandr**
### Description (*)
PR For issue #198 Products: access to related/up-sell/cross-sell product fields
This PR added functionality for query/response with related products.

### Manual testing scenarios (*)
Sample request
```
{
  products(filter: { sku: { eq: "product_dynamic_8" } }) {
    items {
      id
      name
      url_key
      categories {
        id
      }
      crosssell_products {
        items {
          name
          sku
          url_key
          canonical_url
          image {
            url
          }
          price {
            regularPrice {
              amount {
                value
              }
            }
            minimalPrice {
              amount {
                value
              }
            }
            maximalPrice {
              amount {
                value
              }
            }
          }
        }
      }
      upsell_products {
        items {
          name
          sku
          url_key
          canonical_url
          image {
            url
          }
          price {
            regularPrice {
              amount {
                value
              }
            }
            minimalPrice {
              amount {
                value
              }
            }
            maximalPrice {
              amount {
                value
              }
            }
          }
        }
      }
      related_products {
        items {
          name
          sku
          url_key
          canonical_url
          image {
            url
            label
          }
          price {
            regularPrice {
              amount {
                value
                currency
              }
              adjustments {
                amount {
                  value
                  currency
                }
                code
                description
              }
            }
            maximalPrice {
              amount {
                value
                currency
              }
              adjustments {
                amount {
                  value
                  currency
                }
                code
                description
              }
            }
            minimalPrice {
              amount {
                value
                currency
              }
              adjustments {
                amount {
                  value
                  currency
                }
                code
                description
              }
            }
          }
        }
      }
    }
  }
}

```
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/Related/CrossSellProducts.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+use Magento\Framework\GraphQl\Query\ResolverInterface;
+use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
+
+
+class CrossSellProducts implements ResolverInterface
+{
+    /**
+     * Attribute to select fields
+     */
+    public const FIELDS = ['sku', 'name', 'price', 'image', 'url_path', 'url_key'];
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/Related/CrossSellProducts.php:23]: We should take this data from request

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/Related.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<?php
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/Related.php:1]: Pls, add `declare(strict_types=1);` to all of new files
And check code style of all of files

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/Related/CrossSellProducts.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @param \Magento\Framework\GraphQl\Config\Element\Field $field
+     * @param ContextInterface $context
+     * @param ResolveInfo $info
+     * @param array|null $value
+     * @param array|null $args
+     * @throws \Exception
+     * @return mixed|Value
+     */
+    public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/Related/CrossSellProducts.php:45]: We can use `inheritdoc`

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/Related/RelatedProducts.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
+    {
+        $product = $value['model'];
+        $this->dataProvider->addFieldToSelect(self::FIELDS);
+        $collection = $this->dataProvider->getData($product);
+
+        $count = 0;
+        $data = [];
+        foreach ($collection as $item) {
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/Related/RelatedProducts.php:63]: `foreach ($data as $key => $value)`

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Related/AbstractDataProvider.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\Related;
+
+abstract class AbstractDataProvider
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Related/AbstractDataProvider.php:10]: Pls, don't use supertypes only for code reusing (if a relation is not like `type A is subtype of B`)

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Related/CrossSellDataProvider.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * @package Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\Related
+ */
+class CrossSellDataProvider extends AbstractDataProvider
+{
+    /**
+     * @param $product
+     */
+    protected function prepareCollection($product): void
+    {
+        $this->collection = $product->getCrossSellProductCollection();
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Related/CrossSellDataProvider.php:22]: Looks like proxy on another method

File: app/code/Magento/CatalogGraphQl/etc/schema.graphqls [hunk 7]
```diff
@@ -271,6 +271,9 @@ interface ProductInterface @typeResolver(class: "Magento\\CatalogGraphQl\\Model\
     websites: [Website] @doc(description: "An array of websites in which the product is available") @resolver(class: "Magento\\CatalogGraphQl\\Model\\Resolver\\Product\\Websites")
     product_links: [ProductLinksInterface] @doc(description: "An array of ProductLinks objects") @resolver(class: "Magento\\CatalogGraphQl\\Model\\Resolver\\Product\\ProductLinks")
     media_gallery_entries: [MediaGalleryEntry] @doc(description: "An array of MediaGalleryEntry objects") @resolver(class: "Magento\\CatalogGraphQl\\Model\\Resolver\\Product\\MediaGalleryEntries")
+    related_products: RelatedProduct @doc(description: "RelatedProduct") @resolver(class: "Magento\\CatalogGraphQl\\Model\\Resolver\\Product\\Related")
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/etc/schema.graphqls:274]: Looks like resolver should be `Magento\\CatalogGraphQl\\Model\\Resolver\\Product\\Related\\RelatedProducts`
Why do we need one more layer?
The same for another types of related products

**Review by naydav — CHANGES_REQUESTED** [review]

**Comment by naydav** [comment]
Hello @VoronoyAlexandr,
We still need some changes to finish the PR, will you continue progress? 
Thank you

Please let me know if you need any assistance.

**Comment by VitaliyBoyko** [comment]
Closing due to inactivity.

**Pull request closed without merging**